### DISCUSSION TITLE
How to add a logger (Serilog) to Runtime

### DISCUSSION THREAD

**Discussion by davidyu2023** [louthy/language-ext#1204] [Q&A] [answered]
I am venturing into functional Runtime and trying to add Serilog to it. I may need one or more log files. Can you advice how to bake the file paths into the Runtime?

**Comment by bmazzarol** [comment, marked as answer]
The recipe outlined in the wiki has 4 components, 

1. Trait interface, this gives the runtime access to the IO implementation 
2. Interface for the IO itself, in your case it's a way to log.
3. Implementations of the IO to plug into your runtime 
4. A Generic Static class (module/service ), which uses the IO via the trait interface. This is the surface area you expose to develop against. 

I recommend looking through the existing traits exposed by Sys and reading the wiki again. Build something simple and iterate on it.

IMHO serilog would be an implementation detail,  the capability is logging, the backing library that does it can be anything.

**Reply by davidyu2023** [reply]
@bmazzarol I actually did go through all components and was able to add the logger. I also added a runtime for streaming, and they all work fine. I was just trying to see if there is a nice way to pass info from ```appsettings.json``` to the functional runtime. I give it a try again.

**Reply by louthy** [reply]
If you look at the [example `Runtime` in `LanguageExt.Sys`](https://github.com/louthy/language-ext/blob/main/LanguageExt.SysX/Live/Runtime.cs#L29) - you'll see it is constructed with a [`RuntimeEnv` type](https://github.com/louthy/language-ext/blob/main/LanguageExt.SysX/Live/Runtime.cs#L164) that is used to carry in configuration data.  You can either expose this configuration through property accessors (which makes sense for configuration that is shared over multiple types of runtime - live and test, for example); or you can pass it into the constructor of the underlying IO implementation classes when you access their trait members:

For example, you see this a lot:

```c#
        public Eff<Runtime, FileIO> FileEff =>
            SuccessEff(Sys.Live.FileIO.Default);

        public Eff<Runtime, DirectoryIO> DirectoryEff =>
            SuccessEff(Sys.Live.DirectoryIO.Default);

        public Eff<Runtime, TextReadIO> TextReadEff =>
            SuccessEff(Sys.Live.TextReadIO.Default);

       ...
```
Those traits don't need any configuration, so they can simply return a default implementation.  You could do something like this:

```c#
        public Eff<Runtime, LoggingIO> LoggingEff =>
            Eff<Runtime, LoggingIO>(rt => new SerilogIO(rt.Env.SerilogConfig));
```
That will get the settings out of the `RuntimeEnv` and into your `SerilogIO` implementation.  As suggested by @bmazzarol , the `SerilogIO` here is the implementation detail, it gets abstracted away by the `LoggingIO` trait.  Allowing you to plug in different IO solutions, either through configuration, or by defining multiple runtimes.

**Reply by davidyu2023** [reply]
Thanks @louthy this connected the dots for me. I have always wondered what the ```RuntimeEnv``` is about, and should have taken the "Env" part of the name more seriously.
